dev/ctia/dev/split_tests.clj

```diff
@@ -98,6 +101,69 @@
     ;select this split
     (nth all-splits this-split)))
 
+(defn this-split-using-scheduling-with-full-knowledge [timings [this-split total-splits] nsyms]
+  {:pre [(map? timings)
+         (vector? nsyms)
+         (seq nsyms)]
+   :post [(vector? %)]}
+  (let [extra-namespaces (set/difference (set nsyms)
+                                         (set (keys timings)))
+        unallocated-splits (into (priority-map-keyfn (juxt :duration :id))
+                                 (map (fn [split]
+                                        [split {:id split
+                                                :duration 0
+                                                :nsyms []}]))
+                                 (range total-splits))
+        slowest-to-fastest-timings (->> (concat timings
+                                                (map vector
+                                                     extra-namespaces
+                                                     ;; allocated last
+                                                     (repeat {:elapsed-ns 0})))
+                                        (map (fn [[nsym {:keys [elapsed-ns]}]]
+                                               {:pre [(simple-symbol? nsym)
+                                                      (<= 0 elapsed-ns)]}
+                                               [elapsed-ns nsym]))
+                                        (into (sorted-set))
+                                        rseq)
+        ;; algorithm: always allocate new work to the fastest job, and process
+        ;;            work from slowest to fastest
+        splits (reduce (fn [so-far [elapsed-ns nsym :as current-timing]]
```

Comment:
this `reduce` is much easier to read than the previous `loop`. Here we properly identify that we iterate on the ascending timings and accumulate the splits.
thx.

Reply:
Excellent thanks!